File: publication-figure-renderer/python/src/publication_figure_renderer/evagene_client.py
"""Thin client for the two Evagene endpoints this demo needs."""

from __future__ import annotations

from typing import Any

from .http_gateway import HttpGateway

_HTTP_OK_LOWER = 200
_HTTP_OK_UPPER = 300


class ApiError(RuntimeError):
    """Raised when the Evagene API is unreachable or returns a non-2xx response."""


class EvageneClient:
    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        http: HttpGateway,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._http = http

    def fetch_pedigree_svg(self, pedigree_id: str) -> str:
        """Return the raw SVG document for a pedigree."""
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}/export.svg"
        response = self._http.get(
            url, headers=self._headers(accept="image/svg+xml")
        )
        self._require_ok(response, url)
        return response.text

    def fetch_pedigree_detail(self, pedigree_id: str) -> dict[str, Any]:
        """Return the pedigree detail (metadata + resolved individuals)."""
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}"
        response = self._http.get(
            url, headers=self._headers(accept="application/json")
        )
        self._require_ok(response, url)
        try:
            payload = response.json()
        except ValueError as exc:
            raise ApiError(f"Evagene API returned invalid JSON for {url}: {exc}") from exc
        if not isinstance(payload, dict):
            raise ApiError(
                f"Evagene API returned non-object JSON for {url}: "
                f"{type(payload).__name__}"
            )
        return payload

    def _headers(self, *, accept: str) -> dict[str, str]:
        return {"X-API-Key": self._api_key, "Accept": accept}

    @staticmethod
    def _require_ok(response: Any, url: str) -> None:
        if not _HTTP_OK_LOWER <= response.status_code < _HTTP_OK_UPPER:
            raise ApiError(f"Evagene API returned HTTP {response.status_code} for {url}")

File: publication-figure-renderer/r/R/evagene_client.R
# Thin client for the two Evagene endpoints this demo needs.

HTTP_OK_LOWER <- 200L
HTTP_OK_UPPER <- 300L

#' Raise an api_error condition.
#' @noRd
stop_api <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("api_error", "error", "condition")
  ))
}

#' Fetch the SVG export of a pedigree.
#'
#' @param gateway A gateway function (see `make_httr2_gateway()`).
#' @param base_url Evagene base URL (trailing slash optional).
#' @param api_key Value for the `X-API-Key` header.
#' @param pedigree_id UUID string.
#' @return The raw SVG document as a single character string.
fetch_pedigree_svg <- function(gateway, base_url, api_key, pedigree_id) {
  url <- sprintf(
    "%s/api/pedigrees/%s/export.svg", sub("/$", "", base_url), pedigree_id
  )
  response <- gateway(list(
    method = "GET",
    url = url,
    headers = c("X-API-Key" = api_key, "Accept" = "image/svg+xml")
  ))
  require_ok(response, url)
  response$body_text
}

#' Fetch pedigree detail (metadata + resolved individuals/relationships/eggs).
#'
#' @param gateway A gateway function (see `make_httr2_gateway()`).
#' @param base_url Evagene base URL (trailing slash optional).
#' @param api_key Value for the `X-API-Key` header.
#' @param pedigree_id UUID string.
#' @return A named list parsed from the JSON response.
fetch_pedigree_detail <- function(gateway, base_url, api_key, pedigree_id) {
  url <- sprintf(
    "%s/api/pedigrees/%s", sub("/$", "", base_url), pedigree_id
  )
  response <- gateway(list(
    method = "GET",
    url = url,
    headers = c("X-API-Key" = api_key, "Accept" = "application/json")
  ))
  require_ok(response, url)

  parsed <- tryCatch(
    jsonlite::fromJSON(response$body_text, simplifyVector = FALSE),
    error = function(e) {
      stop_api(sprintf("Evagene API returned invalid JSON: %s", conditionMessage(e)))
    }
  )
  if (!is.list(parsed) || is.null(names(parsed))) {
    stop_api("Evagene API returned non-object JSON for pedigree detail")
  }
  parsed
}

require_ok <- function(response, url) {
  if (response$status < HTTP_OK_LOWER || response$status >= HTTP_OK_UPPER) {
    stop_api(sprintf("Evagene API returned HTTP %d for %s", response$status, url))
  }
}
